### PR TITLE
[9.x] Add a new way of defining local scopes on Eloquent models

### PR DESCRIPTION
**Pull request by ambengers**
Hi Team,

Following the changes in how we define Eloquent model accessors and mutators, I am putting forward these proposals on how we can define local scopes as well.

Currently, we are defining local scopes like this:

```
class Ticket extends Model
{
    public class scopePrice($builder, $value)
    {
        $builder->where('price', $value);
    }
}
```
# Proposal 1
Which is currently the changes in this PR - we can define local scopes in the model within the static `$scopes` array, like so:

```
class Ticket extends Model
{
    protected static $scopes = [
        'price' => Price::class,
    ];
}
```

Then, we can use the artisan `make:scope` command to generate a local scope class by adding the `--local` option.

```
$ php artisan make:scope Price --local
```

This will generate the local scope class with an `apply()` method that accepts the Builder instance as the first argument, and `$value` as the second argument for dynamic local scopes.

```
use Illuminate\Database\Eloquent\Builder;

class Price
{
    public function apply(Builder $builder, $value)
    {
        $builder->where('price', $value);
    }
}
```

# Proposal 2
Which I can submit a separate PR for - we can define a `scope()` method in the model class that will return an instance of a local scope class, where the scopes can be defined as public methods.

```
class Ticket extends Model
{
    public function scope($builder) : LocalScope
    {
        return new TicketScope($builder);
    }
}

```

The scope class can then extend an abstract class defined in Laravel to ensure that it will always accept the builder instance as its dependency through the constructor
```
abstract class LocalScope
{
    protected $builder;

    public function __construct(Builder $builder)
    {
        $this->builder = $builder;
    }
}

class TicketScope extends LocalScope
{
    public function price($value)
    {
        $this->builder->where('price', $value);
    }

    // Other named scopes can be define here...
}
```

Like the first proposal, we can use the artisan `make:scope` command to generate a local scope class by adding the `--local` option.

```
$ php artisan make:scope TicketScope --local
```

Or we can have both because it's always good to have options, so why not.

### CODE REVIEW

File: src/Illuminate/Foundation/Console/ScopeMakeCommand.php [hunk 1]
```diff
@@ -73,4 +78,16 @@ protected function getDefaultNamespace($rootNamespace)
     {
         return is_dir(app_path('Models')) ? $rootNamespace.'\\Models\\Scopes' : $rootNamespace.'\Scopes';
```
[the remark this answers is not in the document]
Comment (ambengers) [reply src/Illuminate/Foundation/Console/ScopeMakeCommand.php:79]: @dammy001 that's the default for Laravel when creating global scopes. Not part of my changes for this PR.

**Comment by rodrigopedra** [comment]
Since I learned one could override the creation of the Eloquent Builder for a model, I've been using this technique outlined in this blog post by Tim MacDonald on his blog:

https://timacdonald.me/dedicated-eloquent-model-query-builders

In my opinion, one advantage this technique has over the proposed solution is not to have multiple single purpose classes. Another one, after adding the right docblocks on top of the model's class, IDE autocompletion is straightforward.

**Comment by ambengers** [comment]
@rodrigopedra Appreciate your feedback.

I've been doing that query builder override on some of my projects as well, but that approach does not come without its own issues. 

First, it is not officially documented. New developers could take a while to figure that out unless someone/something leads them to the technique/blog post. Second, although arguably very slim, there is always that risk factor that one of the methods (scopes) in your override class is name-clashing with a method in the builder class after an upgrade in the future.

I personally do not have issues with single-purpose classes. But if it's the community/Laravel team's consensus to avoid them, I can propose another - more of a method-based approach. 

We can define the scope class within a `scope` method in the model, then the scope class can accept the builder instance through the constructor and also hold the scopes as its methods, like so:

```
class Ticket extends Model
{
    protected function scope()
    {
        return TicketScope::class;
    }
}

class TicketScope
{
    protected $builder;

    public function __construct(Builder $builder)
    {
        $this->builder = $builder;
    }

    public function status($value)
    {
        $this->builder->where('status', $value);
    }

    public function isPublished()
    {
        $this->builder->where(...);
    }   
}   
```

@taylorotwell I can PR the above approach should we decide to adopt it instead. Let me know. Cheers!

**Comment by rodrigopedra** [comment]
@ambengers thanks for the feedback. Just to be clear I am not against improving how local scopes are defined, I just found your first proposal more verbose than the current way of doing this.

Also, personally I don't see much difference from your second proposal and the technique of using a custom query builder. But I also agree that technique is also a bit verbose and not intuitive at first.

As an alternative I was thinking in something that more closely resemble the new way to define accessors and mutators:

~~~php
class Ticket extends Model
{
    public function price(): LocalScope
    {
        return new LocalScope(fn ($builder, $value) => $builder->where('price', $value));
    }
}
~~~

Where `LocalScope` (I am not very good at naming things), would be a class similar to the `Attribute` class used to define accessors and mutators. And the return type hint would help identifying the scope methods just as it is done with accessors/mutators.

One added benefit would be the `where` count detection to autowrap a scope inside a nested `where`, as it is done today for local scopes, and which is a caveat of using the custom query builder as noted on the linked blog post.

What do you think?

**Comment by ambengers** [comment]
@rodrigopedra no worries! We agree on improving how local scopes are defined - that's what matters in this discussion.

In my personal experience, I find that local scopes could easily grow bigger than mutators/accessors especially if you're working with datasets with multiple ways of filtering. For one, the `Ticket` model that I've been using as an example here is from an actual project. I currently have 21 local scopes defined for that model (in a builder override class), only 4 accessors, and no mutators. 

So I am fine having the accessors/mutators be defined within the model. But for local scopes, we should have a documented option to clean up the model without needing to override the builder class. So we can either be more verbose using an array of single-purpose classes, or a dedicated `TicketScope` class. Or maybe even both, because why not? 

I honestly thought about your approach as well when I was initially experimenting on this PR. The issue is the scope method still needs to be defined in the model, which doesn't really clean up the model. So it's not much different from `scopeFooBar` approach. 

And I also considered the fact that Laravel already has `make:scope` command for generating global scopes. We might want to take advantage of that for local scopes as well.

**Comment by lkmadushan** [comment]
@ambengers can't we move this scope query functionality to the database query builder?

**Comment by ambengers** [comment]
> @ambengers can't we move this scope query functionality to the database query builder?

@lkmadushan Sorry but I'm not sure how that could be an elegant solution or how that could happen without incurring breaking changes.

**Comment by ambengers** [comment]
Updated PR description to describe the proposed approaches.

**Comment by DarkGhostHunter** [comment]
Personally I think it should be `$builders`, because you're extending the Builder instance instead of creating local scopes. 

I vote for `$builders`, because these are just that, query builders. You shouldn't need to resolve your Query Builder class using the Service Container, and the class can be shared across multiple models without adding more multiple `scopeSomething()` to them (but you can replicate the same using a Trait with added PHPDocs).

Personally I think this is a great way to deprecate the `extend()` method of the Global Scope, separating its behaviour.

```php
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\SoftDeleteBuilder;

public function initializeSoftDeletes()
{
    // ...
    
    $this->addBuilder(SoftDeleteBuilder::class);
}
```

I think it would be a great alternative to use `$scopes` with a class that extends an abstract `ExtendedBuilder` class, because that's is what is doing, extending the builder with new methods.

```php
use Illuminate\Database\Eloquent\Model;
use App\Builders\Storage;

/**
 * @mixin \App\Scopes\Local\StorageScopes
 */
class Image extends Model
{
    protected $builders = [Storage::class];

    // ...
}

use Illuminate\Database\Eloquent\ExtendedBuilder;

class Storage extends ExtendedBuilder
{
    public function whereRemote()
    {
        $this->where('is_remote', true);
    }

    public function whereDisk($disk)
    {
        $this->where('stored_in_disk', $disk);
    }
}
```

This kind of scopes should be added before the Global Scope to avoid being eating by generalization.

1. Builder global macro
2. Local Scope
3. Local Scope Class
4. Global Scope (`extend()`)

**Comment by deleugpn** [comment]
To be honest `tap` is such a superior DX over Eloquent Scopes 🤷‍♂️

**Comment by taylorotwell** [comment]
@deleugpn elaborate with an example? 👀

**Comment by ajcastro** [comment]
maybe it would be good to have a survey to ask the community if they still prefer scope queries running under magic __methods.
sometimes when reading code, finding the source code of scope queries is a pain because it cannot be found by IDE by alt+click (find declaration)... I use vscode btw... I think situation is similar to phpstorm... 
I think the trend is now more on going to better static analysis.. just my 2cents tho...

**Comment by ambengers** [comment]
@DarkGhostHunter 

> Personally I think it should be `$builders`, because you're extending the Builder instance instead of creating local scopes.

Actually, extending the builder instance is not what I am trying to do on this proposal. Scopes are supposed to use the existing builder instance of a model and just add more constraints (where clauses) to it. 

IMHO, we really shouldn't be extending the builder instance at all, to avoid the risk of method name-clashing in future upgrades.

Also, the correct name under the hood is `named scopes`. But, for Proposal 1, I wanted to keep Laravel's convention on lookup tables and just use one word so I used `$scopes`. For Proposal 2, I wanted to follow the convention that is used in the documentation which is `local scope`, so that's what I used for naming the abstract class `LocalScope`.

**Comment by DarkGhostHunter** [comment]
> @DarkGhostHunter 
> 
> 
> 
> > Personally I think it should be `$builders`, because you're extending the Builder instance instead of creating local scopes.
> 
> 
> 
> If inheritance is what you mean by **extending** the builder instance, that is not what I am trying to do on both proposals. Scopes are supposed to use the existing builder instance of a model and just add more constraints (where clauses) to it. 

I get that, but anyway you're effect is the same: adding a method to the builder itself. If you're adding an existing method name, you're gonna have a bad time. There is already an scoping mechanism (two indeed). You can use a trait to remove them from the Model and share them across.

> Proposal 1 accepts the builder instance as the first argument of the `apply` method, just like how the current local scope is implemented. And Proposal 2 will accept the builder instance through the constructor.

I'm not saying these proposals are bad, but rather they're missing the point: extending the builder with new scopes (related to the model being queried, of course). The effect is the same: `$query->whereSomething()`.

> IMHO, we really shouldn't be extending the builder instance at all to avoid the risk of method name-clashing in future upgrades. Eloquent has this powerful feature of forwarding calls to wherever it wants so let's just leverage that.

But using an class with new methods can be accessible with just that. Instance it, save it inside the Eloquent Builder instance, call the method, forward it to the class, profit. My code doesn't extends the Eloquent Builder, but is based on your approach: take out the scopes into its own class.

> Lastly, the correct name under the hood is `named scopes`. But, for Proposal 1, I wanted to follow Laravel's convention on lookup tables and just use one word so I used `$scopes`. For Proposal 2, I wanted to follow the convention that is used in the documentation which is `local scope`, so that's what I used for naming the abstract class `LocalScope`.

I'm bad at naming things. I humbly declare myself as not competent in naming things unless nobody cares.

**Comment by lkmadushan** [comment]
@taylorotwell I guess @deleugpn's idea was this.

```php
class MatchingEmail {

    public function __construct(protected string $email) {}

    public function __invoke($query) 
    {

        $query->where('email', $this->email);
    }

}


User::query()->tap(new MatchingEmail('taylor@laravel.com'))->get();
```

Thanks, @deleugpn your comment helped me to solve the issue of writing a complex scope method.

**Comment by DarkGhostHunter** [comment]
I'm thinking that a better approach should be allowing the Model to register a bunch of Local Scopes from a class (Proposal 2) instead of just one by the class name, but also allow it to be invokable. This way you could make the `published()` in `Article::query()->published()` work be either:

1. Declaring it using `scopePublished()` in the model (takes precedence)
2. Adding it to the scope list in `$scopes`.
3. Using `addLocalScopes(Published::class)` at boot time.

This way a class is not entirely tied to one local scope, but may that may share a common procedure. The developer would have something like this:

```php
class Published extends LocalScopes
{
    public function published()
    {
        $this->where('published_at', '<', now());
    }

    public function pendingPublishing()
    {
        $this->where('published_at', '>', now());
    }
    
    public function notPublished()
    {
        $this->whereNull('published_at');
    }
}
```

The way we can implement this behavior is simple: 

1. When a call to a Builder method that doesn't exists rebounds into the Model, AND it's the local scope `scopeSomething()` doesn't exists,
2. it just then populate the `$scope` array by retrieving all public methods from these classes. 
3. Once done, we check the list if the method exists. 
4. If that's the case, we instantiate the class, put in the Builder instance, call the method and return the result.

Hey, what about a simple invokable class? Well, we can use `is_callable` over the class and call it if  `$method === lcfirst($class)`.

**Comment by taylorotwell** [comment]
> @taylorotwell I guess @deleugpn's idea was this. 👀
> 
> ```
> class MatchingEmail {
> 
>     public function __construct(protected string $email) {}
> 
>     public function __invoke($query) 
>     {
> 
>         $query->where('email', $this->email);
>     }
> 
> }
> 
> 
> User::query()->tap(new MatchingEmail('taylor@laravel.com'))->get();
> ```
> 
> Thanks, @deleugpn your comment helped me to solve the issue of writing a [query object](https://www.martinfowler.com/eaaCatalog/queryObject.html) using Laravel.

Personally I just suggest people do this if they want another way to do "local scope" like behavior.

**Comment by dillingham** [comment]
Never needed it but it might be nice to have apply-able scopes
```php
User::query()->scope(new MatchingEmail('taylor@laravel.com'))->get();
```

**Comment by ambengers** [comment]
> Personally I just suggest people do this if they want another way to do "local scope" like behavior.

@taylorotwell noted on this. 

Personally, I still think chainable methods are still a much cleaner approach, especially when applying multiple scopes.

```
User::matchingEmail('taylor@laravel.com')->verified()->get();
```

Maybe something you'd be interested in going back for the version 10 release, at least?

Cheers!

**Comment by ajcastro** [comment]
For chainable methods, I would prefer to use Tim Mcdonald's approach because of better static analysis. If ever your scope can be shared across different models, it can be easily achieved using traits. 

```php
// App\Models\User.php
use  App\Models\Builders\UserBuilder;

class User extends Model {
    public function newEloquentBuilder($query): UserBuilder
    {
        return new UserBuilder($query);
    }
}

//  App\Models\Builders\UserBuilder.php
use Illuminate\Database\Eloquent\Builder;
use App\Models\Builders\Concerns\CanQueryActive;

class UserBuilder extends Builder
{
    use CanQueryActive;
}

// App\Models\Builders\Concerns\CanQueryActive.php

trait CanQueryActive 
{
    public function activeOnly() 
    {
        $this->where('is_active', 1);
    }
}
```

I think the better action we can do is to update the docs so that we can use the `tap()` on query builder for class-based scopes and override the eloquent builder for the chainable-methods approach, making them an official and accepted approach.
What do you think guys?

**Pull request closed without merging**